src/user.rs

```diff
@@ -424,15 +408,146 @@ async fn on_room_message(
 
 /// This function returns homeserver domain and url, ex:
 ///  - get_homeserver_url("matrix.domain.com") => ("matrix.domain.com", "https://matrix.domain.com")
-fn get_homeserver_url<'a>(homeserver: &'a str, protocol: Option<&'a str>) -> (&'a str, String) {
+fn get_homeserver_url(homeserver: &str, protocol: Option<&str>) -> (String, String) {
     let regex = Regex::new(r"https?://").unwrap();
     if regex.is_match(homeserver) {
         let parts: Vec<&str> = regex.splitn(homeserver, 2).collect();
-        (parts[1], homeserver.to_string())
+        (parts[1].to_string(), homeserver.to_string())
     } else {
         let protocol = protocol.unwrap_or("https");
-        (homeserver, format!("{protocol}://{homeserver}"))
+        (homeserver.to_string(), format!("{protocol}://{homeserver}"))
+    }
+}
+
+async fn get_client(homeserver_url: &str, retry_enabled: bool) -> Option<Client> {
+    let instant = Instant::now();
+
+    let (_, homeserver) = get_homeserver_url(homeserver_url, None);
+
+    log::info!("Attempt to create a client");
+
+    let timeout = Duration::from_secs(30);
+
+    let request_config = if retry_enabled {
+        RequestConfig::new().retry_timeout(timeout)
+    } else {
+        RequestConfig::new().disable_retry().timeout(timeout)
+    };
+
+    let client = Client::builder()
+        .request_config(request_config)
+        .homeserver_url(homeserver)
+        .build()
+        .await;
+    if client.is_err() {
+        println!("got request config error {}", client.err().unwrap());
+        log::info!("Failed to create client");
+        return None;
+    }
+
+    log::info!("New client created {}", instant.elapsed().as_millis());
+
+    Some(client.unwrap())
+}
+
+struct UserParams {
+    server: String,
+    progress_bar: ProgressBar,
+    i: i64,
+    retry_attempts: usize,
+    timestamp: u128,
+    client: Client,
+    tx: Sender<Event>,
+    respect_login_well_known: bool,
+}
+
+fn create_user(user_params: UserParams) -> impl futures::Future<Output = Option<User<Registered>>> {
+    let client_arc = Arc::new(tokio::sync::Mutex::new(user_params.client));
+    let progress_bar = user_params.progress_bar.clone();
+
+    async move {
+        let id = format!("user_{}_{}", user_params.i, user_params.timestamp);
```

Comment:
Do we really need to use the timestamp still? I think this is partly making our lives much harder when trying to come up with the desired users in the `run` method and would be good to avoid it.